### DISCUSSION TITLE
Transfer entity between two worlds

### DISCUSSION THREAD

**Discussion by TerranWorks** [bevyengine/bevy#11101] [Q&A] [answered]
TL;DR I want to be able to losslessly transfer items from the main world to another world (stored as a resource) and back again

I'm implementing an inventory system and the method I've settled on involves moving an entity from the game world to another world and storing some additional metadata in a small database. With this approach I can essentially "freeze" an entity (it doesn't render, get physics updates, collide with anything, etc) while persisting its component data in a form that I can still query using bevy ecs. At the moment, I have an exclusive system which takes in a ``&mut World`` and utilizes ``SystemState`` but I keep getting hung up on not being able to get a mutable reference to World more than once and I can't figure out how I would even begin transfering an entity between worlds. Here is my exclusive system:
```rust
    fn handle_inventory_event (world: &mut World){
        world.resource_scope(|world, mut cached_state: Mut<CachedSystemState>| {
            world.resource_scope(|world, inventory_data: Mut<InventoryData>| {
                for event in cached_state.event_state.get(world).read() { 
                    match event {
                        InventoryEvent::InsertItems(insert_data) => {
                            for insert_item in insert_data {
                                // transfer insert_item.item_id from main world to inventory_data.world
                            }
                        },
                        InventoryEvent::RetrieveItems(retrieve_data) => {},
                        InventoryEvent::DisposeItems(dispose_data) => {},
                        InventoryEvent::TransferItems(transfer_data) => {},
                    }
                }
            });
        });
```

I've seen some references to MultiWorld support but I don't know how far along that is. What would be a good next step?

**Comment by TimJentzsch** [comment]
To my understanding, the multi world stuff is currently special-cased for rendering and not yet available for general use.
In the Bevy 0.12 announcement post, I found a reference to the [`ExtractInstancesPlugin`](https://bevyengine.org/news/bevy-0-12/#usage), but this seems to be specific to the rendering world.

I also found an [RFC on subworlds](https://github.com/bevyengine/rfcs/pull/16) and a [discussion on multi-world support](https://github.com/bevyengine/bevy/discussions/7893).

So I think this is not easily possible yet.
Your best shot might be to take a look at the [render world code](https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/extract_instances.rs) and try to implement something custom for your use-cases.

**Comment by nicopap** [comment, marked as answer]
This is definitively possible, but not the recommended way of approaching the problem you are describing. I would typically add an `Inventory` component, and remove physics and rendering components to "convert" a physical entity into an inventory item. I've done this many times. I think you would benefit from not looking at different `World`s as different "container" where to put entities. But rather entirely ignore the concept of `World` (unless you care about serialization and scene loading, or other esoteric stuff). It's an implementation details, and shouldn't enter in game logic. Use components to "tag" entities with various behaviors.

If you want to "move" entities between  `World`s, it is possible to do so in bevy, but I don't think there is an easy way of only moving a subset or a single entity between worlds. Bevy does this for `Scene` loading. However, you can only transfer components that implement `Reflect` and are registered, furthermore, if the component in question has an `Entity` field, it should also implement `MapEntities`.

**Reply by TerranWorks** [reply]
Adding an InventoryItem component and removing render and physics was the first thing I tried, but it got messy when I wanted to move things from the inventory into the world. This was because the objects have may have different colliders or other component data, which means I need to add different components back for different kinds of objects. At the moment, I'm just adding RigidBodyDisabled and setting its Visibility::Hidden.

Would it be possible to somehow remove the components and store them within the InventoryItem component, so then I can just re-add them when I want to respawn the entity? If not, then I'm sure I can make my current solution work

**Reply by TerranWorks** [reply]
I ended up going with your suggestion so I'll close this discussion
